plugins/services/src/js/utils/ServiceUtil.js

```diff
@@ -135,27 +108,51 @@ const ServiceUtil = {
     return serviceName[serviceName.length - 1];
   },
 
-  convertServiceLabelsToArray(service) {
-    if (!(service instanceof Service)) {
-      return [];
-    }
+  isSDKService(service) {
+    return (
+      service &&
+      service.getLabels &&
+      service.getLabels().DCOS_COMMONS_API_VERSION != null
+    );
+  },
 
-    const labels = service.getLabels();
-    if (labels == null) {
-      return [];
+  /**
+   * this function extracts webURL logic from all structs to keep it in one place
+   *
+   * @param {object} labels object from service
+   * @param {string} rootUrl from config
+   * @returns {string} url or empty string
+   */
+  getWebURL(labels, rootUrl) {
+    const {
+      DCOS_SERVICE_NAME,
+      DCOS_SERVICE_PORT_INDEX,
+      DCOS_SERVICE_SCHEME,
+      DCOS_SERVICE_WEB_PATH,
+      DCOS_COMMONS_API_VERSION
+    } = labels;
+    const serviceName = encodeURIComponent(DCOS_SERVICE_NAME);
+
+    // service name label not set or empty
```

Comment:
as i understood you like stylistic discussions, i'd vote for having a couple of intention-revealing booleans to replace the comments and use them for the early return. pseudo-code:


```typescript
    // ...
    const serviceName = encodeURIComponent(DCOS_SERVICE_NAME);
    const hasNewWebUiLabel = serviceName && DCOS_SERVICE_WEB_PATH;
    const hasOldWebUiLabel = serviceName && DCOS_SERVICE_PORT_INDEX && DCOS_SERVICE_SCHEME && !DCOS_COMMONS_API_VERSION;

    if (hasNewWebUiLabel){ // i'd prefer a nested ternary, but eslint does not like that.
      return `${rootUrl}/service/${serviceName}${DCOS_SERVICE_WEB_PATH}` // <-- happy path or new/best/most common way of doing things at the point of most attention.
    } else if (hasOldWebUiLabel) {
      return `${rootUrl}/service/${serviceName}/`;
    } else {
      return "";
    } 
```

or something along those lines...

Reply:
yeah, thanks for that - i just shuffled this code around too much to see where I can further improve it, like your suggestion 👍